### PR TITLE
Publish allennlp-server to PyPI and misc CI changes

### PR DESCRIPTION
**Pull request by agururajvais**
Changes for Issue: https://github.com/allenai/allennlp/issues/4974

Acknowledgment of support by @pab-vmware , @prajaktakini-vmware, @annajung, @divijbajaj

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,56 @@
...
+        <img alt="Build" src="https://github.com/allenai/allennlp-server/workflows/CI/badge.svg?event=push&branch=main">
+    </a>
+    <a href="https://pypi.org/project/allennlp-server/">
+        <img alt="PyPI" src="https://img.shields.io/pypi/v/allennlp-server">
+    </a>
+    <a href="https://github.com/agururajvais/allennlp-server/blob/master/LICENSE">
+        <img alt="License" src="https://img.shields.io/github/license/allenai/allennlp-server?color=blue&cachedrop">
+    </a>
+    <a href="https://codecov.io/gh/allenai/allennlp">
+        <img alt="Codecov" src="https://codecov.io/gh/allenai/allennlp/branch/main/graph/badge.svg">
```
Comment (pab-vmware) [inline README.md:23]: Misleading unless you can point to the allenlp-server instance itself.

File: setup.py [hunk 2]
```diff
@@ -1,5 +1,73 @@
...
+        "License :: OSI Approved :: Apache Software License",
+        "Programming Language :: Python",
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3 :: Only",
+        "Programming Language :: Python :: 3.6",
+        "Programming Language :: Python :: 3.7",
+        "Programming Language :: Python :: 3.8",
+        "Topic :: Scientific/Engineering",
+        "Topic :: Scientific/Engineering :: Artificial Intelligence",
+        "Topic :: Scientific/Engineering :: Visualization",
```
Comment (pab-vmware) [inline setup.py:44]: Not sure if this is applicable. 
Probably Dirk can confirm.

File: README.md [hunk 3]
```diff
@@ -1,8 +1,56 @@
...
+
+### From PyPI
+
+`allennlp-server` is available on PyPI. To install with `pip`, just run
+
+```bash
+pip install allennlp-server
+```
+
+Note that the `allennlp-server` package is tied to the [`allennlp` core package](https://pypi.org/projects/allennlp-server). Therefore when you install the models package you will get the corresponding version of `allennlp` and `allennlp-models` (if you haven't already installed `allennlp`). For example,
```
Comment (pab-vmware) [inline README.md:43]: It requires both allennlp core and the models package

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -76,4 +193,34 @@ jobs:
...
+      - name: Download server package
+        uses: actions/download-artifact@v1
+        with:
+          name: server-package
+          path: dist
+
+      - name: Upload to PyPI
+        env:
+          PYPI_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
+        run: twine upload -u allennlp -p $PYPI_PASSWORD dist/*
```
Comment (pab-vmware) [inline .github/workflows/main.yml:226]: github indicates a missing EOF newline

**Comment by gururajvaishnav** [comment]
> Changes for Issue: [allenai/allennlp#4974](https://github.com/allenai/allennlp/issues/4974)

Acknowledgment of support by` @pab-vmware, https://github.com/prajaktakini-vmware https://github.com/annajung https://github.com/divijbajaj `

**Comment by gururajvaishnav** [comment]
@dirkgr I am contributing as VMware and need to get approval before merging, and is currently in process of getting the repo approved. Can you review and if all is good, can we wait for a couple of days before we can merge the changes? :D

**Comment by agururajvais** [comment]
@dirkgr  I am contributing as VMware and need to get approval before merging, as it is currently in process of getting my contribution approved. Can you review and if all is good, can we wait for a couple of days before we can merge the changes? :D

**Comment by gururajvaishnav** [comment]
> Changes for Issue: [allenai/allennlp#4974](https://github.com/allenai/allennlp/issues/4974)

Acknowledgment of support by @pab-vmware, @prajaktakini-vmware @annajung @divijbajaj

File: setup.py (hunk 2 above)
Comment (dirkgr) [reply setup.py:44]: I don't think we can say this. This doesn't really visualize the output (though we keep talking about doing that at some point ...).

**Review by dirkgr — COMMENTED** [review]
This looks good to me, but I'd rather have @epwalsh take a look. He did the publishing pipeline for the rest of AllenNLP.

File: .github/dependabot.yml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+version: 2
+updates:
+- package-ecosystem: pip
+  directory: "/"
+  schedule:
+    interval: daily
+    time: "13:00"
+  open-pull-requests-limit: 10
+  ignore:
+  - dependency-name: nr-databind-core
```
Comment (epwalsh) [inline .github/dependabot.yml:10]: Why is this needed?

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -59,11 +59,128 @@ jobs:
       run: |
         pytest -v
 
+    - name: Upload coverage to Codecov
+      if: matrix.python == '3.7' && github.repository == 'allenai/allennlp-server' && (github.event_name == 'push' || github.event_name == 'pull_request')
```
Comment (epwalsh) [inline .github/workflows/main.yml:63]:
```suggestion
      if: github.repository == 'allenai/allennlp-server' && (github.event_name == 'push' || github.event_name == 'pull_request')
```

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -59,11 +59,128 @@ jobs:
...
+    if: github.repository == 'allenai/allennlp-server'
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Setup Python
+        uses: actions/setup-python@v1
+        with:
+          python-version: 3.7
```
Comment (epwalsh) [inline .github/workflows/main.yml:89]: Let's be consistent with 3.8:

```suggestion
          python-version: 3.8
```

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -59,11 +59,128 @@ jobs:
...
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Cleanup directory
+        run: |
+          # Remove the source code so that it doesn't conflict with the wheel
+          # installation.
+          rm -rf allennlp_server/
+
```
Comment (epwalsh) [inline .github/workflows/main.yml:159]: How about just not checking out the code in the first place?

```suggestion
```

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -77,3 +194,33 @@ jobs:
...
+    needs: [checks, build_package, test_package, docker]
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Setup Python
+        uses: actions/setup-python@v1
+        with:
+          python-version: 3.7
```
Comment (epwalsh) [inline .github/workflows/main.yml:211]:
```suggestion
          python-version: 3.8
```

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -77,3 +194,33 @@ jobs:
...
+  # Publish the distribution files to PyPI
+  publish:
+    name: PyPI
+    if: github.repository == 'allenai/allennlp-server' && (github.event_name == 'release' || github.event_name == 'schedule')
+    needs: [checks, build_package, test_package, docker]
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+
```
Comment (epwalsh) [inline .github/workflows/main.yml:207]: Probably don't need to checkout here either.

```suggestion
```

**Review by epwalsh — CHANGES_REQUESTED** [review]
Hey @agururajvais, this looks good, thanks for taking this on. Just a few comments.

Oh, and can you remove the `setup.cfg` file?

File: setup.py (hunk 2 above)
Comment (agururajvais) [reply setup.py:44]: Cool. Will remove this.

File: .github/dependabot.yml (hunk 5 above)
Comment (agururajvais) [reply .github/dependabot.yml:10]: The whole of dependabot or just nr-databind-core? I used the same config from allennlp to keep things consistent.

Comment (epwalsh) [reply .github/dependabot.yml:10]: Ah, gotcha. We definitely want dependabot, so thanks for adding that. But I don't think we need this `ignore` section for `nr-databind-core`.

File: README.md [hunk 11]
```diff
@@ -1,8 +1,51 @@
...
+
+### From PyPI
+
+`allennlp-server` is available on PyPI. To install with `pip`, just run
+
+```bash
+pip install allennlp-server
+```
+
+Note that the `allennlp-server` package is tied to the [`allennlp` core package](https://pypi.org/projects/allennlp) and [`allennlp-models` package](https://pypi.org/projects/allennlp-models). Therefore when you install the models package you will get the corresponding version of `allennlp` and `allennlp-models` (if you haven't already installed `allennlp` or `allennlp-models`). For example,
```
Comment (epwalsh) [inline README.md:38]:
```suggestion
Note that the `allennlp-server` package is tied to the [`allennlp` core package](https://pypi.org/projects/allennlp) and [`allennlp-models` package](https://pypi.org/projects/allennlp-models). Therefore when you install the server package you will get the latest compatible version of `allennlp` and `allennlp-models` (if you haven't already installed `allennlp` or `allennlp-models`). For example,
```

**Review by epwalsh — APPROVED** [review]
LGTM! Thanks so much!

I'll trigger a release tomorrow so we get this on PyPI.

**Pull request merged**